STYLE-GUIDE.md

```diff
@@ -0,0 +1,285 @@
+
+## Formatting
+
+2 spaces everywhere. Tabs are forbidden. Haskell indentation can be unpredictable, so generally stick with what your editor suggests.
+There is no hard line-length limit, though if you go beyond 110 or 120 you should generally split it up, especially for type signatures.
+
+### Use applicative notation when constructing simple data types
+
+``` haskell
+thing :: Parser Foo
+
+-- Broke:
+thing = do
+  a <- bar
+  b <- baz
+  pure (Foo a b)
+
+-- Woke:
+thing = Foo <$> bar <*> baz
+```
+
+Code written with applicative notation is less powerful than code written in `do`-notation.  As a result, it can be easier to reason about, as long as it's also readable.
+
+Overreliance on applicative notation can create code that is difficult to read. Don't use applicative notation in combination with operator sections. If in doubt, write it with `do` notation and see if it's more immediately comprehensible.
+
+### Use leading commas for records, exports, and lists
+
+Leading commas make it easy to add and remove fields without introducing syntax errors, and properly aligned records are easy to read:
+
+``` haskell
+data Pos = Pos
+  { posLine   :: Int
+  , posColumn :: Int
+  }
+```
+
+This rule is enforced by CI.
+
+## Naming
+
+Locally bound variables (such as the arguments to functions, or helpers defined in a `where` clause) can have short names, such as `x` or `go`. Globally bound functions and variables should have descriptive names.
+
+You'll often find yourself implementing functions that conflict with Prelude definitions. If this is the case, avoid adding a prefix to these functions, and instead import them qualified.
+
+``` haskell
+-- Bad
+foo = heapLookup thing
+-- Better
+foo = Heap.lookup thing
+```
+
+## Functions
+
+### Don't go crazy with infix operators
+
+Sensible use of infix operators can provide serious readability benefits, but often the best tool is just a named function. If you're defining new operators, make sure that you have a solid justification for doing so.
+
+### Avoid list comprehensions
+
+In almost all cases, `map`, `filter`, `fold`, and the `[]` monad are more flexible and readable.
+
+### Don't go crazy with point-free definitions
+
+Point-free style can help or hinder readability, depending on the context. If a function is expressed naturally with the `.` operator, then do so, but if you have to over-complicate the definition to write it point-free, then you should probably just write out the variable names. If you are reviewing someone else's PR and find a point-free definition hard to read, ask them to simplify/clarify it.
+
+### Prefer `.` and `$` to parentheses
+
+Parentheses can make a function harder to edit, since parentheses have to be balanced. The composition and application operators (`.` and `$`) can reduce clunkiness.
+
+``` haskell
+-- Bad
+f (g (h x))
+-- Better
+f $ g $ h x
+-- Hlint will reject this
```

Comment:
I'm confused at this comment/example

Reply:
I re-read it, not sure why you're confused.  Can you elaborate?

Reply:
I'm confused why hlint will reject it, or why we're showing this example after "Better". If the style we prefer is the mix of `.` and `$` -- e.g.  `f . g . h $ i x` -- I'd expect to see that example here instead

Reply:
Ah, yeah, that makes sense.  I'll rework it.

Reply:
@cnr I reworked the example, is that better?

Reply:
I'd kill the last example that hlint rejects -- I think it just adds confusion